Moving and deleting multiple trees in skeletontracing

MichaelBuessemeyer opened the pull request:
This PR enables selecting multiple trees. Moving and deleting effect all selected trees.

### URL of deployed dev instance (used for testing):
- [https://movinganddeletingmultipletrees.webknossos.xyz](https://movinganddeletingmultipletrees.webknossos.xyz)

### Steps to test:
- open a skeletontracing
- create some trees and groups
- select some of them with ctrl + left mouse
- try deleting and moving the trees
- there should be no possibility to have a parent and a child selected simultaneously
- when deleting and groups are selected a modal should show up that asks whether to remove the groups and the subtrees or only the groups


### Issues:
- fixes #3143

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- ~~[ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~~
- [x] Updated [documentation](../blob/master/docs) if applicable -> maybe
- ~~[ ] Needs datastore update after deployment~~
- [x] Ready for review

MichaelBuessemeyer commented:
TODO: 
- [ ] discuss keys that are used to select a tree (there might be a better short than alt + left mouse)
- [ ]  discuss how to better deselect everything
- [ ] no tooltip at mouse position possible while dragging trees/groups -> other possibilities?

philippotto commented on app/assets/javascripts/oxalis/view/right-menu/delete-groups-modal-view.js:
```diff
@@ -0,0 +1,41 @@
+  onCancel: () => void,
+};
+
+export default function DeleteGroupsModalView({
```
Very idiomatic code :+1:

philippotto commented on app/assets/javascripts/oxalis/view/right-menu/tree_hierarchy_view.js:
```diff
@@ -130,12 +135,28 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
     const treeId = evt.target.dataset.id;
-    this.props.onSetActiveTree(parseInt(treeId, 10));
+    if (evt.ctrlKey) {
+      this.props.handleTreeSelect(parseInt(treeId, 10));
```
I'd just change line 137 to `const treeId = parseInt(evt.target.dataset.id, 10)` to avoid duplication. Then you can just pass `treeId` here and in the else branch.

philippotto commented on app/assets/javascripts/oxalis/view/right-menu/tree_hierarchy_view.js:
```diff
@@ -147,24 +168,70 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
     }));
   };
 
+  // rekursivly sets the parent of all currently selected groups to the given newParent
```
recursively

philippotto commented on app/assets/javascripts/oxalis/view/right-menu/tree_hierarchy_view.js:
```diff
@@ -210,6 +262,25 @@ class TreeHierarchyView extends React.PureComponent<Props, State> {
 
+  getNodeStyleClassForBackground = (id, isGroup = false) => {
+    if (isGroup) {
+      if (this.props.selectedTreeGroups.includes(id) && this.props.activeGroupId === id) {
```
Please write the results of the `includes` calls into local variables to avoid redundant computations.

philippotto commented on docs/tracing_ui.md:
```diff
@@ -119,8 +119,8 @@ All operations and information regarding trees are organized under a tab called
-Trees can be dragged and dropped between tree groups.
-However over existing tree groups the bring up a little menu for creating new groups, renaming, and deletion.
+Trees can be dragged and dropped between tree groups. This action can be applied to multiple trees by selecting them with Ctrl + Left Mouse. 
+However over existing tree groups to bring up a little menu for creating new groups, renaming, and deletion.
```
`Hover` instead of `However`

philippotto commented:
I didn't review tree_hierarchy_view.js and trees_tab_view.js.

MichaelBuessemeyer commented:
@daniel-wer I agree with you that removing the group selecting will make the ui a lot easier and also everything would be more clear. 
@philippotto @daniel-wer But when I remove the group selection feature, how should I handle dragging a group. Should this only move the group or should all selected trees be moved too? In other words, when should shall all selected tree be moved? Only when a tree is moved or also when a group is moved?

Second: Now that can't have selected groups anymore should I take the group deletion feature out of the "delete" button in the trees tab? (wasn't previously available)

MichaelBuessemeyer commented:
@philippotto  @daniel-wer  This PR is ready for another review :) (200 line of code less :tada: )

daniel-wer commented on app/assets/javascripts/oxalis/view/right-menu/trees_tab_view.js:
```diff
@@ -303,10 +446,33 @@ class TreesTabView extends React.PureComponent<Props, State> {
+        <div className="tree-hierarchy-header">
+          <span>{selectionInfo}</span>
+          <br />
+          <span className="clickable-text" onClick={this.deselectEverything}>
```
My design proposal:

![tree-selected](https://user-images.githubusercontent.com/1702075/49385975-da660a00-f71e-11e8-9ba1-046d3f5be3ba.png)

Code roughly like this:
```javascript
<Alert
  type="info"
  style={{ marginRight: "auto" }}
  message={
    <React.Fragment>
      2 Tree(s) selected.{" "}
      <Button type="dashed" size="small">
        Clear Selection
      </Button>
    </React.Fragment>
  }
/>
```

daniel-wer commented:
Just for the record, the action items from our call:
- [ ] Only show the "x trees are selected" message if selectedTrees.length > 1
- [ ] Selecting another tree (without ctrl) or group will clear selection (as in file picker)
- [ ] Active Tree is always selected, if multiple trees are selected all should be green
- [ ] If a group is active, no other trees can be selected with ctrl + click (maybe show message)

MichaelBuessemeyer commented:
@daniel-wer I applied your feedback: You can now only have eighter selected trees, an active group or an active tree.

daniel-wer commented on app/assets/javascripts/oxalis/view/right-menu/trees_tab_view.js:
```diff
@@ -114,8 +131,74 @@ class TreesTabView extends React.PureComponent<Props, State> {
+      const subtrees = groupToTreesMap[groupId] != null ? groupToTreesMap[groupId] : [];
+      if (withTrees) {
+        // also delete all subtrees
+        this.props.onDeleteMultipleTrees(subtrees.map(tree => tree.treeId));
```
My question was rather what would happen to group 2 and tree 1 when the user removes group 1 and selects `Remove group and subtrees`?

daniel-wer replied:
I just tested this and in the current version this leads to inconsistent state. group2 and tree1 are no longer present in the tree tab, but tree1 is still existing in the tracing. The tree and group deletion needs to be done recursively when the user selected this option.

daniel-wer commented on app/assets/javascripts/oxalis/view/right-menu/trees_tab_view.js:
```diff
@@ -318,6 +334,13 @@ class TreesTabView extends React.PureComponent<Props, State> {
 
+  deleteGroupAndHideModal(groupToDelete: ?number, deleteSubtrees = false) {
+    this.hideDeleteGroupsModal();
+    if (groupToDelete !== null && groupToDelete !== undefined) {
```
`if (groupToDelete != null) {` is equivalent

daniel-wer commented on docs/tracing_ui.md:
```diff
@@ -120,7 +120,7 @@ A typical skeleton annotation consists of one or more trees.
 Tree groups can have a name and are used to structure and label your annotation even further.
 Trees can be dragged and dropped between tree groups. This action can be applied to multiple trees by selecting them with Ctrl + Left Mouse. 
-Hover over existing tree groups to bring up a little menu for creating new groups, renaming, and deletion.
+Hover over existing tree groups to bring up a little menu for creating new groups and deletion. Renaming of a group can be done my selecting a group and then entering the new name into the input above the tree hierarchy structure view.
```
by instead of my

MichaelBuessemeyer commented:
@daniel-wer I applied your feedback: I fixed the error with the inconsistent state after deleting a group with all subtrees. And I added better handling for the corner cases of the tree selection.

philippotto commented:
> I would prefer the second option, because I can imagine this functionality is useful, what's your opinion on this @philippotto?

Yes, that makes perfect sense. The two options should be either deleting a minimum or a maximum of reasonable trees. Deleting all children recursively appears to be the more frequent scenario. The caption of the confirm dialog needs to be adapted then, as well ("Do want to delete this group and keep its children or do you want to remove all children recursively?" or something like that).

daniel-wer commented:
@MichaelBuessemeyer Awesome, works as expected for me :)

@philippotto Maybe you could do a test run of the functionality as well, this PR has been open for a while and I want to avoid "Betriebsblindheit" :)

If that is successful, ship it :ship:

philippotto commented:
I just tested this branch and it works pretty well :+1: The UI is very intuitive in my opinion and easy to grasp.

However, I noticed one edge case: If there is only a group at the root level (which contains all other groups and trees), deleting that group recursively will result in a completely empty tracing. By itself, this sounds fine, however, the rest of WK assumes that there will always be at least one (empty) tree. In the other places we ensure that by adding an empty tree if the last tree was added (see the skeleton reducer for details). This would need to be fixed before merging, I think :)

daniel-wer commented:
Without explicitly debugging this, but from reading the code, I think the bug is a little more subtle than this. The `deleteMultipleTreesAsUserAction` _does_ use the usual `deleteTreeAction`, so when deleting the last tree, a new tree is automatically created. However, the new tree is assigned to the group of the active tree (the last tree), which is probably one of the deleted groups, which is then deleted afterwards.
I'll think about a possible solution.
